### PR TITLE
sha256 alignment fixes/rework

### PR DESCRIPTION
**Pull request by poettering**
Prompted by: https://github.com/systemd/systemd/pull/25636#issuecomment-1339978702

### CODE REVIEW

File: src/fundamental/macro-fundamental.h [hunk 1]
```diff
@@ -327,14 +327,23 @@ static inline size_t ALIGN_TO(size_t l, size_t ali) {
...
 #ifndef SD_BOOT
 /* libefi also provides ALIGN, and we do not use them in sd-boot explicitly. */
 #define ALIGN(l)  ALIGN_TO(l, sizeof(void*))
 #define ALIGN_PTR(p) ((void*) ALIGN((uintptr_t) (p)))
 #endif
 
+/* Checks if the specified pointer is aligned as appropriate for the specific type */
+#define ALIGNED16_P(p) (((uintptr_t) p) % __alignof__(uint16_t) == 0)
+#define ALIGNED32_P(p) (((uintptr_t) p) % __alignof__(uint32_t) == 0)
+#define ALIGNED64_P(p) (((uintptr_t) p) % __alignof__(uint64_t) == 0)
```
Comment (medhefgo) [inline src/fundamental/macro-fundamental.h]: Why not `_PTR` for consistency?

Comment (medhefgo) [reply src/fundamental/macro-fundamental.h]: Also, these don't do the aligning, but the name suggests they do. Should really be called `IS_ALIGNED` or something like that.

**Review by yuwata — APPROVED** [review]
LGTM, except for the naming, as suggested in the above.

**Review by keszybz — COMMENTED** [review]
LGTM too except for the naming.

File: src/fundamental/macro-fundamental.h (hunk 1 above)
Comment (poettering) [reply src/fundamental/macro-fundamental.h]: Oh, young grasshopper! The "_P" suffix stands for "predicate". It's a lisp thing. Don't they teach that anymore? ;-)

Basically, in lisp you suffix functions that return boolean with -p, to indicate they are a "predicate". Some of the older folks also do this in C, including glibc internally at various places. This macro is derived from the glibc internal one (which existed in the sha256.c code beforehand, see the diff), and I figured we could just stick to that.

See this:

https://jblevins.org/log/def-predicate_p
https://www.gnu.org/software/emacs/manual/html_node/elisp/Coding-Conventions.html
https://stackoverflow.com/questions/70545047/what-does-suffix-p-mean-in-common-lisp-functions

Some C APIs do this too, for example

**Comment by poettering** [comment]
Renamed the macro to a boring ISALIGNED32(), accepting that you youngsters don't know your good ol' LISP anymore. ;-)

File: src/fundamental/macro-fundamental.h [hunk 2]
```diff
@@ -327,14 +327,23 @@ static inline size_t ALIGN_TO(size_t l, size_t ali) {
...
 #ifndef SD_BOOT
 /* libefi also provides ALIGN, and we do not use them in sd-boot explicitly. */
 #define ALIGN(l)  ALIGN_TO(l, sizeof(void*))
 #define ALIGN_PTR(p) ((void*) ALIGN((uintptr_t) (p)))
 #endif
 
+/* Checks if the specified pointer is aligned as appropriate for the specific type */
+#define ISALIGNED16(p) (((uintptr_t) p) % __alignof__(uint16_t) == 0)
+#define ISALIGNED32(p) (((uintptr_t) p) % __alignof__(uint32_t) == 0)
+#define ISALIGNED64(p) (((uintptr_t) p) % __alignof__(uint64_t) == 0)
```
Comment (keszybz) [inline src/fundamental/macro-fundamental.h]: Hmm, why not `IS_ALIGNED16`? We have various `is_*` functions, and they are spelled with the underscore. (We also have `_PTR` above, and it is inconsistent to have no underscore here.)

File: src/fundamental/macro-fundamental.h [hunk 3]
```diff
@@ -327,14 +327,23 @@ static inline size_t ALIGN_TO(size_t l, size_t ali) {
         return ((l + ali - 1) & ~(ali - 1));
 }
 
+#define ALIGN2(l) ALIGN_TO(l, 2)
 #define ALIGN4(l) ALIGN_TO(l, 4)
 #define ALIGN8(l) ALIGN_TO(l, 8)
+#define ALIGN2_PTR(l) ((void*) ALIGN_TO((uintptr_t) l, 2))
+#define ALIGN4_PTR(l) ((void*) ALIGN_TO((uintptr_t) l, 4))
+#define ALIGN8_PTR(l) ((void*) ALIGN_TO((uintptr_t) l, 8))
```
Comment (keszybz) [inline src/fundamental/macro-fundamental.h]:
```C
#define ALIGN2_PTR(l) ((void*) ALIGN2((uintptr_t) l)
...
```
?
This makes it clearer that this is just about casting the types back and forth.

**Comment by medhefgo** [comment]
> Renamed the macro to a boring ISALIGNED32(), accepting that you youngsters don't know your good ol' LISP anymore. ;-)

I know *of* lisp. And there is only so many brackets a mortal can handle before going insane.

Also, I believe in learning from history and its many mistakes it buries.

File: src/fundamental/macro-fundamental.h (hunk 2 above)
Comment (poettering) [reply src/fundamental/macro-fundamental.h]: well, there's also ISPOWEROF2 and similar. we have it one way, we have it the other way, too.

this is bikeshedding ultimately. but I am fine with either, hence to cut this short, will change to IS_ now

**Comment by poettering** [comment]
force pushed a new version with the suggested renames.

**Review by keszybz — APPROVED** [review]
LGTM.

(What would systemd be without the eternal style bikeshedding?)

**Comment by medhefgo** [comment]
> (What would systemd be without the eternal style bikeshedding?)

One could lesson our own load by training an AI to do that for us… 🤔

**Pull request merged**